##### Contents:
###### 1. Import Libraries
###### 2. Import data
###### 3. Data consistency checks
###### 4. Mixed Type Data
###### 4.1 Missing Values
###### 4.2 Duplicate Values
###### 5. Task
###### 6. Exporting dataframe

#### 01. Importing Libraries

In [3]:
#Import libraries
import pandas as pd
import numpy as np
import os

#### 02. Importing data

In [17]:
#Create path
path = r'C:\Users\Daniella\ACH4 - Instacart Basket Analysis'
#Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02. Data', 'Original Data', 'products.csv'), index_col = False)
# Import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02. Data', 'Prepared Data', 'orders_wrangled(new).csv'), index_col = False)

#### 03. Data Consistency Checks

In [19]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_prior_order  
count            3.421083e+06       3.421083e+06            3.214874e+06  
mean             2.776219e+00       1.345202e+01            1.111484e+01  
std              2.046829e+00       4.226088e+00            9.206737e+00  
min              0.000000e+00       0.000000e+00            0.000000e+00  
25%              1.000000e+00       1.000000e+01            4.000000e+00  
50%              3.000000e+00       1.300000e+01            7.000000e+00  
75%              5.000000e+00       1.600000e+01            1.500000e+01  
max              6.000000e+00       2.300000e+01            3.000000e+01

#### 04. Mixed Type Data

In [22]:
# Create a dataframe
df_test = pd.DataFrame()

In [24]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [26]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [28]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


#### 4.1 Missing Values

In [35]:
# Function to obtain the total number of missing observations
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [118]:
# Create a subset containing only missing value
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [120]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [43]:
df_prods.shape

(49693, 5)

In [45]:
# Create a subset containing only non-missing values:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [47]:
df_prods_clean.shape

(49677, 5)

In [122]:
##Drop all missing values
df_prods.dropna(inplace = True)

In [124]:
##drop only the NaNs from a particular column
df_prods.dropna(subset = ['product_name'], inplace = True)

#### 4.2 Duplicate Values

In [126]:
# Looking for full duplicates within the df_prods_clean:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [128]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [130]:
df_prods_clean.shape

(49677, 5)

In [132]:
# Dropping the duplicate values by creating a new dataframe:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [134]:
df_prods_clean_no_dups.shape

(49672, 5)

#### 05. Task

In [136]:
# 2. Check df_ords dataframe for inconsistencies
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_prior_order  
count            3.421083e+06       3.421083e+06            3.214874e+06  
mean             2.776219e+00       1.345202e+01            1.111484e+01  
std              2.046829e+00       4.226088e+00            9.206737e+00  
min              0.000000e+00       0.000000e+00            0.000000e+00  
25%              1.000000e+00       1.000000e+01            4.000000e+00  
50%              3.000000e+00       1.300000e+01            7.000000e+00  
75%              5.000000e+00       1.600000e+01            1.500000e+01  
max              6.000000e+00       2.300000e+01            3.000000e+01

##### Answer: The count of days_since_prior_order has a lower count that the other columns. This could mean that there are missing values in the days_since_prior_order column.

In [138]:
# 3. Check for mixed-type data in df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

##### Answer: No mixed-type data found in df_ords dataframe

In [139]:
# 5. Run a check for missing values in your df_ords dataframe:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##### Answer: The days_since_prior_order coulmn has a total of 206,209 mssing values. These missing values could be the first orders placed by customers.

In [142]:
# 6. How to handle missing values:
# Create a subset containing only missing values
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [144]:
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0                 0   2539329        1             1                       2   
11               11   2168274        2             1                       2   
26               26   1374495        3             1                       1   
39               39   3343014        4             1                       6   
45               45   2717275        5             1                       3   
...             ...       ...      ...           ...                     ...   
3420930     3420930    969311   206205             1                       4   
3420934     3420934   3189322   206206             1                       3   
3421002     3421002   2166133   206207             1                       6   
3421019     3421019   2227043   206208             1                       1   
3421069     3421069   3154581   206209             1                       3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [146]:
##Check the frequency of order_number=1 in df_nan
df_nan['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

In [148]:
# Check the frequency of order_number=1 in df_ords
df_ords['order_number'].value_counts()

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

In [150]:
# Count of missing values in 'days_since_prior_order'
missing_count = df_ords['days_since_prior_order'].isnull().sum()
print(missing_count)

206209


In [152]:
# Count of orders where order_number is 1
first_orders_count = df_ords[df_ords['order_number'] == 1].shape[0]
print(first_orders_count)

206209


##### Answer: The fact that the number of missing values (206,209) in days_since_prior_order is the same as the number of first orders (order_number=1) suggests that the missing values are likely for the first orders placed. 

In [156]:
# 7. Run a check for duplicate values in df_ords dataframe
df_ords_dups = df_ords[df_ords.duplicated()]

In [158]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_prior_order]
Index: []

#### 06. Exporting Dataframes

In [99]:
# Export df_prods as a CSV file 
df_prods_clean_no_dups.to_csv(os.path.join(path, '02. Data', 'Prepared Data', 'products_checked(new).csv'))

In [105]:
# Exporting df_ords dataframe as \"orders_checked.csv
df_ords.to_csv(os.path.join(path, '02. Data', 'Prepared Data', 'orders_checked(new).csv'))